File: python/sorting/Quicksort.py
def swap(a, b, arr):
    if a!=b :
        arr[a], arr[b] = arr[b], arr[a]

def partition(elements, start , end):
    pivot_index = start
    pivot = elements[pivot_index]

    while (start < end):

        while start < len(elements) and elements[start] <= pivot :
            start +=1

        while elements[end] > pivot:
            end -=1
        
        if start < end:
            swap(start, end, elements)
        
    swap(pivot_index,end, elements)

    return end

def quick_sort(elements, start ,end):
    if start < end:
        pi = partition(elements,start ,end)
        quick_sort(elements, start, pi -1) # left portion
        quick_sort(elements, pi+1, end) # right partition

if __name__ == "__main__":
    ele = [11, 45, 29, 1, 6, 35]
    quick_sort(ele, 0, len(ele)-1)
    print(ele)

File: c/sorting/quickSort.c
/* Quick sort */

#include<stdio.h>
void swap(int *num1, int *num2) {
    int temp = *num1;
    *num1 = *num2;
    *num2 = temp;
}
int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);
    for(int j = low; j < high; j++) {
        if(arr[j] < pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}
void quickSort(int arr[], int low, int high) {
    if(low < high) {
        int p = partition(arr, low, high);
        quickSort(arr, low, p-1);
        quickSort(arr, p+1, high);
    }
}
int main(void) {
    int arr[] = {6, 4, 9, 2, 8};
    int size = sizeof(arr) / sizeof(arr[0]);
    quickSort(arr, 0, size-1);
    for(int i =0; i<size; i++) {
        printf("%d\t", arr[i]);
    }
}
/*
-----output------
2       4       6       8       9 
*/